### PR TITLE
Added multiple reducers support for team-level parallel reduce

### LINKED ISSUE #4510: Multiple reductions for team-level loops
I'd like to be able to use multiple reductions for parallel_reduce calls on TeamThreadRange/TeamVectorRange/ThreadVectorRange levels.

### PR DESCRIPTION
**Pull request by ldh4**
Resolves #4510 

Added an interface to allow parallel_reduce with TeamThreadRange, ThreadVectorRange or TeamVectorRange policy to be called with more than one reducer.

### CODE REVIEW

**Comment by ldh4** [comment]
Retest this please.

**Comment by ldh4** [comment]
Retest this please.

File: core/src/impl/Kokkos_Combined_Reducer.hpp
Comment (masterleinad) [inline core/src/impl/Kokkos_Combined_Reducer.hpp:518]: Should this rather be a `team_barrier`?

File: core/src/impl/Kokkos_Combined_Reducer.hpp
Comment (masterleinad) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:506]:
```suggestion
  using mem_space_type = typename MemberType::execution_space::memory_space;
```

Comment (dalg24) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:506]: How comes no test failed?

Comment (masterleinad) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:506]: For the outer-level `parallel_reduce`, we decide whether we need to call deep_copy or whether we can directly write the reduced value back. AFAICT, this doesn't apply to the nested one since we know that we can directly write the reduced value.

Comment (ldh4) [reply core/src/impl/Kokkos_Combined_Reducer.hpp]: Precisely what @masterleinad described. `parallel_reduce` called from the host code needs to consider in which memory space the final reduced value resides before a deep_copy back to the host memory, but the team `parallel_reduce` implementation in this PR expects to be called only from the device code with one of the team policies: `TeamThreadRange`, `ThreadVectorRange` or `TeamVectorRange`.
So, as long as the combined reducers use a consistent device memory space internally, reduction on device side shouldn't be affected.

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (ldh4) [reply core/unit_test/TestTeamCombinedReducers.hpp]: Applied in [d65b5f3](https://github.com/kokkos/kokkos/pull/5727/commits/d65b5f3c6f4f20f540b91ade4aaa55a1e8704685)

File: core/src/impl/Kokkos_Combined_Reducer.hpp
Comment (ldh4) [reply core/src/impl/Kokkos_Combined_Reducer.hpp]: Converted to a `static` function in [a28463e](https://github.com/kokkos/kokkos/pull/5727/commits/a28463eaae89518a32405a9df9e9cb0394bb3d55).
I recall needing to remove `static` from that function at one point, because something was unable to work with a device static function at that time. But that may no longer be the case now. The CI will confirm this.

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (tcclevenger) [inline core/unit_test/TestTeamCombinedReducers.hpp:429]: There is a lot of repeated lines with this identical message in each of the 3 test cases. Would it make sense to just have one test case with the 9 function calls, since they also all use the same struct? Just a thought, feel free to ignore.

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (tcclevenger) [inline core/unit_test/TestTeamCombinedReducers.hpp:432]: Would there be any value in testing also with base case `n=0` for at least one of the cases?

File: core/src/impl/Kokkos_Combined_Reducer.hpp
Comment (masterleinad) [inline core/src/impl/Kokkos_Combined_Reducer.hpp:213]: Where is this used?

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (ldh4) [reply core/unit_test/TestTeamCombinedReducers.hpp:450]: Good point. I added in base case of `n=0` for each team policy and found some deficiencies in these tests.
It turns out that when `n=0` and when the inner `parallel_reduce` do not run, the outer `parallel_for` loop still runs, initializing intermediate team reduced values based on the corresponding reducer.
Added in if conditions to avoid unnecessary aggregates when the inner `parallel_reduce` do not run.

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (ldh4) [reply core/unit_test/TestTeamCombinedReducers.hpp:447]: I agree that these lines are quite redundant across the 3 test cases, but I think there is some value in keeping these unit tests separate based on their team policy. If any of these tests fails, I think I would find it simpler to tell which team policy failed based on the failing test name, whereas it may not be as clear at first glance if all 9 cases were in a single test.

File: core/src/impl/Kokkos_Combined_Reducer.hpp
Comment (ldh4) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:213]: It's used in where each reducer's value is initialized and passed into provided functor.
For example, for `Cuda`:
https://github.com/kokkos/kokkos/blob/15989dd3a1edb72bbc178fc74d081502e32dd3d6/core/src/Cuda/Kokkos_Cuda_Team.hpp#L630-L649

File: core/src/impl/Kokkos_Combined_Reducer.hpp
Comment (ldh4) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:518]: I'll resume on this shortly.
I recall looking into a workaround to consistently call `team_barrier` for all cases after learning that a reference to `team_member` is not stored in `ThreadVectorBoundariesStruct` for a few backends because of their constructors that do not require `team_member`.

Comment (ldh4) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:518]: @masterleinad I've been giving rounds of thoughts on this, and I'm thinking that the fence operations are not necessary here, both fences on line 518 and 524.

These fences were to ensure that reduced values are ready to be copied back to the original references for all reducers in a combined reducer. This is meaningful if reductions are done on device and if the result from device needs to be copied back to reducers on host (this is what the regular combined reducer does).
However, for team combined reducers, both combined reduction part and copying back to its reducer part are on device and are done in sequence.

For example, for cuda `team_reduce` that is called from `parallel_reduce` that takes in a team policy:
https://github.com/kokkos/kokkos/blob/c09dd1c94861541b2f9651474ecdbec9e074dab7/core/src/Cuda/Kokkos_Cuda_Team.hpp#L192-L204
Syncing threads is already done within the reduction. Then the result from (combined) reduction is assigned in (combined) reducer, and thus another fence afterward shouldn't be necessary to start referencing reduced values on device.

Comment (masterleinad) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:518]: Sounds good!

Comment (ldh4) [reply core/src/impl/Kokkos_Combined_Reducer.hpp]: Removed fences in [af212c0](https://github.com/kokkos/kokkos/pull/5727/commits/af212c0bc124056dc22706f18b1ac3da5c06f7ef).

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

**Review by Rombur — APPROVED** [review]

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (dalg24) [inline core/unit_test/TestTeamCombinedReducers.hpp:55]: No we don't want to guard this.

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (dalg24) [inline core/unit_test/TestTeamCombinedReducers.hpp:42]: Do not initialize, we need to make sure it is being handled properly when reducing over an empty range.

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (ldh4) [reply core/unit_test/TestTeamCombinedReducers.hpp:66]: No particular reason. Just wanted to test something more than just 2 reducers.

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (ldh4) [reply core/unit_test/TestTeamCombinedReducers.hpp]: Changed in [39b91e8](https://github.com/kokkos/kokkos/pull/5727/commits/39b91e8209eba18cecb3d077e97c85c6c7a9d2dc).

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (ldh4) [reply core/unit_test/TestTeamCombinedReducers.hpp]: Changed in [39b91e8](https://github.com/kokkos/kokkos/pull/5727/commits/39b91e8209eba18cecb3d077e97c85c6c7a9d2dc).

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (ldh4) [reply core/unit_test/TestTeamCombinedReducers.hpp]: Removed in [39b91e8](https://github.com/kokkos/kokkos/pull/5727/commits/39b91e8209eba18cecb3d077e97c85c6c7a9d2dc).

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (ldh4) [reply core/unit_test/TestTeamCombinedReducers.hpp]: I don't think there was a particular reason. Changed to assigns in [39b91e8](https://github.com/kokkos/kokkos/pull/5727/commits/39b91e8209eba18cecb3d077e97c85c6c7a9d2dc).

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (dalg24) [inline core/unit_test/TestTeamCombinedReducers.hpp:392]: Technically they all should be reduction_identity if n is equal to 0
```suggestion
    if (n == 0) {
      EXPECT_EQ(Kokkos::reduction_identity<int>::sum(), hostView(0));
      EXPECT_EQ(Kokkos::reduction_identity<int>::sum(), hostView(1));
      EXPECT_EQ(Kokkos::reduction_identity<int>::max(), hostView(2));
      EXPECT_EQ(Kokkos::reduction_identity<int>::sum(), hostView(3));
    } else {
      EXPECT_EQ((n * (n + 1) / 2), hostView(0));
      EXPECT_EQ((n * (n + 1) / 2), hostView(1));
      EXPECT_EQ(n, hostView(2));
      EXPECT_EQ(n * n, hostView(3));
    }
```
Same comment everywhere.

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (dalg24) [inline core/unit_test/TestTeamCombinedReducers.hpp:20]:
```suggestion
#include <gtest/gtest.h>
```

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (dalg24) [inline core/unit_test/TestTeamCombinedReducers.hpp:22]:
```suggestion
namespace {
```

**Review by dalg24 — COMMENTED** [review]
Please explicitly compare against reduction identity when the range is empty.
Looks good other than that.

File: core/unit_test/TestTeamCombinedReducers.hpp
Comment (ldh4) [reply core/unit_test/TestTeamCombinedReducers.hpp]: Changed in [2b7a12b](https://github.com/kokkos/kokkos/pull/5727/commits/2b7a12b285fce804e0f299e22aa59ce332397810).

**Comment by dalg24** [comment]
I plan on ignoring 
```
5: [ RUN      ] cuda.workgraph_fib
5: /var/jenkins/workspace/Kokkos/core/unit_test/TestWorkGraph.hpp:127: Failure
5: Expected equality of these values:
5:   h_values(0)
5:     Which is: 5778
5:   full_fibonacci(m_input)
5:     Which is: 6765
5: [  FAILED  ] cuda.workgraph_fib (1426 ms)
```
in the CUDA-11.6-NVCC build.  Still waiting on at least one of the HIP build to pass before I merge.

**Comment by ldh4** [comment]
Retest this please.

**Pull request merged**
